File: PMEditor/src/nl/adaptivity/diagram/android/AndroidCanvas.java
package nl.adaptivity.diagram.android;

import nl.adaptivity.diagram.Rectangle;
import nl.adaptivity.diagram.Theme;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.Path;
import android.graphics.RectF;


public class AndroidCanvas implements IAndroidCanvas {

  private class OffsetCanvas implements IAndroidCanvas {
    /** The offset of the canvas. This is in scaled coordinates. */
    private double aXOffset;
    private double aYOffset;
    private double aScale;


    public OffsetCanvas(OffsetCanvas pBase, Rectangle pArea, double pScale) {
      aXOffset = (pBase.aXOffset - pArea.left)*pScale;
      aYOffset = (pBase.aYOffset - pArea.top)*pScale;
      aScale = pBase.aScale*pScale;
    }

    public OffsetCanvas(OffsetCanvas pBase, double pScale) {
      aXOffset = (pBase.aXOffset)*pScale;
      aYOffset = (pBase.aYOffset)*pScale;
      aScale = pBase.aScale*pScale;
    }

    public OffsetCanvas(Rectangle pArea, double pScale) {
      aXOffset = -pArea.left* pScale;
      aYOffset = -pArea.top* pScale;
      aScale = pScale;
    }

    public OffsetCanvas(double pScale) {
      aXOffset = 0;
      aYOffset = 0;
      aScale = pScale;
    }

    @Override
    public IAndroidCanvas childCanvas(Rectangle pArea, double pScale) {
      return new OffsetCanvas(this, pArea, pScale);
    }

    @Override
    public IAndroidCanvas scale(double pScale) {
      return new OffsetCanvas(this, pScale);
    }


    private AndroidPen scalePen(AndroidPen pPen) {
      return pPen.scale(aScale);
    }

    @Override
    public void drawCircle(double pX, double pY, double pRadius, AndroidPen pPen) {
      AndroidCanvas.this.drawCircle(transformX(pX), transformY(pY), pRadius*aScale, scalePen(pPen));
    }

    @Override
    public void drawBitmap(double pLeft, double pTop, Bitmap pBitmap, AndroidPen pPen) {
      AndroidCanvas.this.drawBitmap(transformX(pLeft), transformY(pTop), pBitmap, scalePen(pPen));
    }

    @Override
    public void drawFilledCircle(double pX, double pY, double pRadius, AndroidPen pPen) {
      AndroidCanvas.this.drawFilledCircle(transformX(pX), transformY(pY), pRadius*aScale, scalePen(pPen));
    }

    @Override
    public void drawRect(Rectangle pRect, AndroidPen pPen) {
      AndroidCanvas.this.drawRect(pRect.offsetScaled(-aXOffset, -aYOffset, aScale), scalePen(pPen));
    }

    @Override
    public void drawFilledRect(Rectangle pRect, AndroidPen pPen) {
      AndroidCanvas.this.drawFilledRect(pRect.offsetScaled(-aXOffset, -aYOffset, aScale), scalePen(pPen));
    }

    @Override
    public void drawPoly(double[] pPoints, AndroidPen pPen) {
      AndroidCanvas.this.drawPoly(transform(pPoints), scalePen(pPen));
    }

    @Override
    public void drawFilledPoly(double[] pPoints, AndroidPen pPen) {
      AndroidCanvas.this.drawFilledPoly(transform(pPoints), scalePen(pPen));
    }

    private double[] transform(double[] pPoints) {
      double[] result = new double[pPoints.length];
      final int len = pPoints.length-1;
      for(int i=0; i<len;++i) {
        result[i] = transformX(pPoints[i]);
        ++i;
        result[i] = transformY(pPoints[i]);
      }
      return result;
    }

    public double transformX(double x) {
      return (x-aXOffset)*aScale;
    }

    public double transformY(double y) {
      return (y-aYOffset)*aScale;
    }

    @Override
    public void drawPath(AndroidPath pPath, AndroidPen pStroke, AndroidPen pFill) {
      Path transformedPath = transformPath(pPath);
      if (pFill!=null) {
        AndroidCanvas.this.drawFilledPath(transformedPath, pFill.getPaint());
      }
      if (pStroke!=null) {
        AndroidCanvas.this.drawPath(transformedPath, scalePen(pStroke).getPaint());
      }
    }

    private Path transformPath(AndroidPath pPath) {
      Path transformedPath = new Path(pPath.getPath());
      Matrix matrix = new Matrix();
      matrix.setScale((float)aScale, (float)aScale);
      matrix.preTranslate((float)-aXOffset, (float) -aYOffset);
      transformedPath.transform(matrix);
      return transformedPath;
    }

    @Override
    public void drawRoundRect(Rectangle pRect, double pRx, double pRy, AndroidPen pPen) {
      AndroidCanvas.this.drawRoundRect(pRect.offsetScaled(-aXOffset, -aYOffset, aScale), pRx*aScale, pRy*aScale, scalePen(pPen));
    }

    @Override
    public void drawFilledRoundRect(Rectangle pRect, double pRx, double pRy, AndroidPen pPen) {
      AndroidCanvas.this.drawFilledRoundRect(pRect.offsetScaled(-aXOffset, -aYOffset, aScale), pRx*aScale, pRy*aScale, scalePen(pPen));
    }

    @Override
    public AndroidStrategy getStrategy() {
      return AndroidStrategy.INSTANCE;
    }

    @Override
    public Theme<AndroidStrategy, AndroidPen, AndroidPath> getTheme() {
      return AndroidCanvas.this.getTheme();
    }

    @Override
    public void drawText(TextPos pTextPos, double pLeft, double pBottom, String pText, double pFoldWidth, AndroidPen pPen) {
      AndroidCanvas.this.drawText(pTextPos, transformX(pLeft), transformY(pBottom), pText, pFoldWidth*aScale, scalePen(pPen), aScale);
    }

  }

  private android.graphics.Canvas aCanvas;
  private Theme<AndroidStrategy, AndroidPen, AndroidPath> aTheme;
  
//Only for debug purposes
//  private Paint aRedPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
//  private Paint aGreenPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

  public AndroidCanvas(android.graphics.Canvas pCanvas, Theme<AndroidStrategy, AndroidPen, AndroidPath> pTheme) {
    aCanvas = pCanvas;
    aTheme = pTheme;
//Only for debug purposes
//    aRedPaint.setColor(Color.rgb(255, 0, 0)); aRedPaint.setStyle(Style.FILL);
//    aGreenPaint.setColor(Color.rgb(0, 255, 0)); aGreenPaint.setStyle(Style.FILL);
  }

  @Override
  public IAndroidCanvas childCanvas(Rectangle pArea, double pScale) {
    return new OffsetCanvas(pArea, pScale);
  }

  @Override
  public void drawFilledCircle(double pX, double pY, double pRadius, AndroidPen pPen) {
    Paint paint = pPen.getPaint();
    Style oldStyle = paint.getStyle();
    paint.setStyle(Paint.Style.FILL);
    aCanvas.drawCircle((float) pX, (float) pY, (float) pRadius, paint);
    paint.setStyle(oldStyle);
  }

  @Override
  public void drawCircle(double pX, double pY, double pRadius, AndroidPen pPen) {
    aCanvas.drawCircle((float) pX, (float) pY, (float) pRadius, pPen.getPaint());
  }

  @Override
  public void drawFilledRoundRect(Rectangle pRect, double pRx, double pRy, AndroidPen pPen) {
    Paint paint = pPen.getPaint();
    Style oldStyle = paint.getStyle();
    paint.setStyle(Paint.Style.FILL);
    aCanvas.drawRoundRect(toRectF(pRect), (float)pRx, (float)pRy, pPen.getPaint());
    paint.setStyle(oldStyle);
  }

  @Override
  public void drawRoundRect(Rectangle pRect, double pRx, double pRy, AndroidPen pPen) {
    aCanvas.drawRoundRect(toRectF(pRect), (float)pRx, (float)pRy, pPen.getPaint());
  }

  @Override
  public void drawFilledRect(Rectangle pRect, AndroidPen pPen) {
    Paint paint = pPen.getPaint();
    Style oldStyle = paint.getStyle();
    paint.setStyle(Paint.Style.FILL);
    aCanvas.drawRect(toRectF(pRect), pPen.getPaint());
    paint.setStyle(oldStyle);
  }

  @Override
  public void drawRect(Rectangle pRect, AndroidPen pPen) {
    aCanvas.drawRect(toRectF(pRect), pPen.getPaint());
  }

  @Override
  public void drawPoly(double[] pPoints, AndroidPen pPen) {
    aCanvas.drawPath(toPath(pPoints), pPen.getPaint());
  }

  @Override
  public void drawFilledPoly(double[] pPoints, AndroidPen pPen) {
    Paint paint = pPen.getPaint();
    Style oldStyle = paint.getStyle();
    paint.setStyle(Paint.Style.FILL);
    aCanvas.drawPath(toPath(pPoints), pPen.getPaint());
    paint.setStyle(oldStyle);
  }

  @Override
  public void drawPath(AndroidPath pPath, AndroidPen pStroke, AndroidPen pFill) {
    if (pFill!=null)
      drawFilledPath(pPath.getPath(), pFill.getPaint());
    if (pStroke!=null)
      drawPath(pPath.getPath(), pStroke.getPaint());
  }

  void drawPath(Path path, Paint paint) {
    aCanvas.drawPath(path, paint);
  }

  private void drawFilledPath(Path path, Paint paint) {
    Style oldStyle = paint.getStyle();
    paint.setStyle(Paint.Style.FILL);
    aCanvas.drawPath(path, paint);
    paint.setStyle(oldStyle);
  }

  private static Path toPath(double[] pPoints) {
    Path result = new Path();

    final int len = pPoints.length - 1;
    if (len>0) {
      result.moveTo((float)pPoints[0], (float)pPoints[1]);
      for(int i=2; i<len; ++i) {
        result.lineTo((float)pPoints[i], (float) pPoints[++i]);
      }
      result.close();
    }
    return result;
  }

  private static RectF toRectF(Rectangle pRect) {
    return new RectF(pRect.leftf(), pRect.topf(), pRect.rightf(), pRect.bottomf());
  }

  @Override
  public AndroidStrategy getStrategy() {
    return AndroidStrategy.INSTANCE;
  }

  @Override
  public Theme<AndroidStrategy, AndroidPen, AndroidPath> getTheme() {
    if (aTheme==null) { aTheme = new AndroidTheme(getStrategy()); }
    return aTheme;
  }

  public void setCanvas(Canvas pCanvas) {
    aCanvas = pCanvas;
  }

  @Override
  public IAndroidCanvas scale(double pScale) {
    return new OffsetCanvas(pScale);
  }

  @Override
  public void drawBitmap(double pLeft, double pTop, Bitmap pBitmap, AndroidPen pPen) {
    aCanvas.drawBitmap(pBitmap, (float) pLeft, (float) pTop, pPen.getPaint());
  }

  @Override
  public void drawText(TextPos pTextPos, double pX, double pY, String pText, double pFoldWidth, AndroidPen pPen) {
    drawText(pTextPos, pX, pY, pText, pFoldWidth, pPen, 1);
  }
  
  private void drawText(TextPos pTextPos, double pX, double pY, String pText, double pFoldWidth, AndroidPen pPen, double pScale) {
    final Paint paint = pPen.getPaint();
    paint.setStyle(Style.FILL);
    float left = getLeft(pTextPos, pX, pText, pFoldWidth, pPen, pScale);
    float baseline = getBaseLine(pTextPos, pY, pPen, pScale);
    aCanvas.drawText(pText, left, baseline, paint);
//Only for debug purposes
//    aCanvas.drawCircle(left, baseline, 3f, aRedPaint);
//    aCanvas.drawCircle((float)pX, (float)pY, 3f, aGreenPaint);
  }

  private static float getBaseLine(TextPos pTextPos, double pY, AndroidPen pPen, double pScale) {
    switch (pTextPos) {
    case TOPLEFT:
    case TOP:
    case TOPRIGHT:
      return (float) (pY+(pPen.getTextAscent()*pScale));
    case LEFT:
    case MIDDLE:
    case RIGHT:
      return (float) (pY+(0.5*pPen.getTextAscent()-0.5*pPen.getTextDescent())*pScale);
    case BASELINEMIDDLE:
    case BASELINERIGHT:
    case BASELINELEFT:
      return (float) pY;
    case BOTTOMLEFT:
    case BOTTOMRIGHT:
    case BOTTOM:
      return (float) (pY-(pPen.getTextDescent()*pScale));
    }
    throw new IllegalArgumentException(pTextPos.toString());
  }

  private static float getLeft(TextPos pTextPos, double pX, String pText, double pFoldWidth, AndroidPen pPen, double pScale) {
    switch (pTextPos) {
    case BASELINELEFT:
    case BOTTOMLEFT:
    case LEFT:
    case TOPLEFT:
      return (float) pX;
    case TOP:
    case BASELINEMIDDLE:
    case MIDDLE:
    case BOTTOM:
      return (float) (pX - ((pPen.measureTextWidth(pText, pFoldWidth)*pScale)/2));
    case TOPRIGHT:
    case RIGHT:
    case BASELINERIGHT:
    case BOTTOMRIGHT:
      return (float) (pX - ((pPen.measureTextWidth(pText, pFoldWidth)*pScale)));
    }
    throw new IllegalArgumentException(pTextPos.toString());
  }


}

File: PMEditor/src/main/java/nl/adaptivity/diagram/android/AndroidCanvas.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.diagram.android

import android.graphics.*
import android.graphics.Paint.Style
import nl.adaptivity.diagram.Canvas
import nl.adaptivity.diagram.Pen
import nl.adaptivity.diagram.Rectangle
import nl.adaptivity.diagram.Theme
import org.jetbrains.annotations.Contract


class AndroidCanvas(private var canvas: android.graphics.Canvas,
                    private var _theme: Theme<AndroidStrategy, AndroidPen, AndroidPath>?)
    : IAndroidCanvas {

    override val strategy: AndroidStrategy
        get() = AndroidStrategy.INSTANCE

    override val theme: Theme<AndroidStrategy, AndroidPen, AndroidPath>
        get() {
            return _theme ?: AndroidTheme(strategy).also { _theme = it }
        }

    fun setCanvas(canvas: android.graphics.Canvas) {
        this.canvas = canvas
    }

    override fun childCanvas(offsetX: Double, offsetY: Double, scale: Double): IAndroidCanvas {
        return OffsetCanvas(offsetX, offsetY, scale)
    }

    override fun drawFilledCircle(x: Double, y: Double, radius: Double, fill: AndroidPen) {
        val paint = fill.paint
        val oldStyle = paint.style
        paint.style = Paint.Style.FILL
        canvas.drawCircle(x.toFloat(), y.toFloat(), radius.toFloat(), paint)
        paint.style = oldStyle
    }

    override fun drawCircle(x: Double, y: Double, radius: Double, stroke: AndroidPen) {
        canvas.drawCircle(x.toFloat(), y.toFloat(), radius.toFloat(), stroke.paint)
    }

    override fun drawCircle(x: Double,
                            y: Double,
                            radius: Double,
                            stroke: AndroidPen?,
                            fill: AndroidPen?) {
        if (fill != null) {
            drawFilledCircle(x, y, radius, fill)
        }
        if (stroke != null) {
            drawCircle(x, y, radius, stroke)
        }
    }

    override fun drawFilledRoundRect(rect: Rectangle,
                                     rx: Double,
                                     ry: Double,
                                     fill: AndroidPen) {
        val paint = fill.paint
        val oldStyle = paint.style
        paint.style = Paint.Style.FILL
        canvas.drawRoundRect(toRectF(rect), rx.toFloat(), ry.toFloat(), fill.paint)
        paint.style = oldStyle
    }

    override fun drawRoundRect(rect: Rectangle,
                               rx: Double,
                               ry: Double,
                               stroke: AndroidPen) {
        canvas.drawRoundRect(toRectF(rect), rx.toFloat(), ry.toFloat(), stroke.paint)
    }

    override fun drawRoundRect(rect: Rectangle,
                               rx: Double,
                               ry: Double,
                               stroke: AndroidPen?,
                               fill: AndroidPen?) {
        if (fill != null) {
            drawFilledRoundRect(rect, rx, ry, fill)
        }
        if (stroke != null) {
            drawRoundRect(rect, rx, ry, stroke)
        }
    }

    override fun drawFilledRect(rect: Rectangle, fill: AndroidPen) {
        val paint = fill.paint
        val oldStyle = paint.style
        paint.style = Paint.Style.FILL
        canvas.drawRect(toRectF(rect), fill.paint)
        paint.style = oldStyle
    }

    override fun drawRect(rect: Rectangle, stroke: AndroidPen) {
        canvas.drawRect(toRectF(rect), stroke.paint)
    }

    override fun drawRect(rect: Rectangle,
                          stroke: AndroidPen?,
                          fill: AndroidPen?) {
        if (fill != null) {
            drawFilledRect(rect, fill)
        }
        if (stroke != null) {
            drawRect(rect, stroke)
        }
    }

    override fun drawPoly(points: DoubleArray,
                          stroke: AndroidPen?,
                          fill: AndroidPen?) {

        val androidPath = toPath(points)
        if (fill != null) {
            val fillPaint = fill.paint
            val oldStyle = fillPaint.style
            fillPaint.style = Paint.Style.FILL
            canvas.drawPath(androidPath, fill.paint)
            fillPaint.style = oldStyle
        }
        if (stroke != null) {
            canvas.drawPath(androidPath, stroke.paint)
        }
    }

    override fun drawPath(path: AndroidPath,
                          stroke: AndroidPen?,
                          fill: AndroidPen?) {
        if (fill != null) drawFilledPath(path.path, fill.paint)
        if (stroke != null) drawPath(path.path, stroke.paint)
    }

    internal fun drawPath(path: Path, paint: Paint) {
        canvas.drawPath(path, paint)
    }

    private fun drawFilledPath(path: Path, paint: Paint) {
        val oldStyle = paint.style
        paint.style = Paint.Style.FILL
        canvas.drawPath(path, paint)
        paint.style = oldStyle
    }

    private fun toPath(points: DoubleArray): Path {
        val result = Path()

        val len = points.size - 1
        if (len > 0) {
            result.moveTo(points[0].toFloat(), points[1].toFloat())
            var i = 2
            while (i < len) {
                result.lineTo(points[i].toFloat(), points[++i].toFloat())
                ++i
            }
            result.close()
        }
        return result
    }

    private fun toRectF(rect: Rectangle): RectF {
        return RectF(rect.leftf, rect.topf, rect.rightf, rect.bottomf)
    }


    override fun scale(scale: Double): IAndroidCanvas {
        return OffsetCanvas(scale)
    }

    override fun translate(dx: Double, dy: Double): IAndroidCanvas {
        return if (dx == 0.0 && dy == 0.0) {
            this
        } else OffsetCanvas(dx, dy, 1.0)
    }

    override fun drawBitmap(left: Double, top: Double, bitmap: Bitmap, pen: AndroidPen) {
        canvas.drawBitmap(bitmap, left.toFloat(), top.toFloat(), pen.paint)
    }

    override fun drawText(textPos: Canvas.TextPos,
                          left: Double,
                          baselineY: Double,
                          text: String,
                          foldWidth: Double,
                          pen: AndroidPen) {
        drawText(textPos, left, baselineY, text, foldWidth, pen, 1.0)
    }

    private fun drawText(textPos: Canvas.TextPos,
                         x: Double,
                         y: Double,
                         text: String,
                         foldWidth: Double,
                         pen: AndroidPen,
                         scale: Double) {
        val paint = pen.paint
        paint.style = Style.FILL
        val left = getLeft(textPos, x, text, foldWidth, pen, scale)
        val baseline = getBaseLine(textPos, y, pen, scale)
        canvas.drawText(text, left, baseline, paint)
        //Only for debug purposes
        //    mCanvas.drawCircle(left, baseline, 3f, mRedPaint);
        //    mCanvas.drawCircle((float)pX, (float)pY, 3f, mGreenPaint);
    }

    private fun getBaseLine(textPos: Canvas.TextPos,
                            y: Double,
                            pen: Pen<*>,
                            scale: Double): Float {
        when (textPos) {
            Canvas.TextPos.MAXTOPLEFT,
            Canvas.TextPos.MAXTOP,
            Canvas.TextPos.MAXTOPRIGHT  -> return (y + pen.textMaxAscent * scale).toFloat()

            Canvas.TextPos.ASCENTLEFT,
            Canvas.TextPos.ASCENT,
            Canvas.TextPos.ASCENTRIGHT  -> return (y + pen.textAscent * scale).toFloat()

            Canvas.TextPos.LEFT,
            Canvas.TextPos.MIDDLE,
            Canvas.TextPos.RIGHT        -> return (y + (0.5 * pen.textAscent - 0.5 * pen.textDescent) * scale).toFloat()

            Canvas.TextPos.BASELINEMIDDLE,
            Canvas.TextPos.BASELINERIGHT,
            Canvas.TextPos.BASELINELEFT -> return y.toFloat()

            Canvas.TextPos.BOTTOMLEFT,
            Canvas.TextPos.BOTTOMRIGHT,
            Canvas.TextPos.BOTTOM       -> return (y - pen.textMaxDescent * scale).toFloat()

            Canvas.TextPos.DESCENTLEFT,
            Canvas.TextPos.DESCENTRIGHT,
            Canvas.TextPos.DESCENT      -> return (y - pen.textDescent * scale).toFloat()
        }
    }

    private fun getLeft(textPos: Canvas.TextPos,
                        x: Double,
                        text: String,
                        foldWidth: Double,
                        pen: AndroidPen,
                        scale: Double): Float {
        when (textPos) {
            Canvas.TextPos.BASELINELEFT,
            Canvas.TextPos.BOTTOMLEFT,
            Canvas.TextPos.LEFT,
            Canvas.TextPos.MAXTOPLEFT,
            Canvas.TextPos.ASCENTLEFT  -> return x.toFloat()

            Canvas.TextPos.ASCENT,
            Canvas.TextPos.DESCENT,
            Canvas.TextPos.BASELINEMIDDLE,
            Canvas.TextPos.MAXTOP,
            Canvas.TextPos.MIDDLE,
            Canvas.TextPos.BOTTOM      -> return (x - pen.measureTextWidth(text, foldWidth) * scale / 2).toFloat()

            Canvas.TextPos.MAXTOPRIGHT,
            Canvas.TextPos.ASCENTRIGHT,
            Canvas.TextPos.DESCENTLEFT,
            Canvas.TextPos.DESCENTRIGHT,
            Canvas.TextPos.RIGHT,
            Canvas.TextPos.BASELINERIGHT,
            Canvas.TextPos.BOTTOMRIGHT -> return (x - pen.measureTextWidth(text, foldWidth) * scale).toFloat()
        }
    }


    private inner class OffsetCanvas(xOffset: Double, yOffset: Double, private val scale: Double) : IAndroidCanvas {

        /** The offset of the canvas. This is in scaled coordinates.  */
        private val xOffset: Double = -xOffset
        private val yOffset: Double = -yOffset

        override val strategy: AndroidStrategy
            get() = AndroidStrategy.INSTANCE


        override val theme: Theme<AndroidStrategy, AndroidPen, AndroidPath>
            get() = this@AndroidCanvas.theme


        constructor(base: OffsetCanvas, offsetX: Double, offsetY: Double, scale: Double) :
            this((base.xOffset - offsetX) * scale,
                 (base.yOffset - offsetY) * scale,
                 base.scale * scale)

        constructor(base: OffsetCanvas, scale: Double) :
            this(base.xOffset * scale, base.yOffset * scale, base.scale * scale)

        constructor(scale: Double) : this(0.0, 0.0, scale)

        override fun scale(scale: Double): IAndroidCanvas {
            return OffsetCanvas(this, scale)
        }

        override fun childCanvas(offsetX: Double, offsetY: Double, scale: Double): IAndroidCanvas {
            return OffsetCanvas(this, offsetX, offsetY, scale)
        }


        override fun translate(dx: Double, dy: Double): IAndroidCanvas {
            return OffsetCanvas(xOffset - dx, yOffset - dy, scale)
        }

        @Contract("null -> null; !null -> !null")
        private fun scalePen(pen: AndroidPen?): AndroidPen? {
            return pen?.scale(scale)
        }

        override fun drawCircle(x: Double, y: Double, radius: Double, stroke: AndroidPen) {
            this@AndroidCanvas.drawCircle(transformX(x), transformY(y), radius * scale, scalePen(stroke)!!)
        }

        override fun drawFilledCircle(x: Double,
                                      y: Double,
                                      radius: Double,
                                      fill: AndroidPen) {
            this@AndroidCanvas.drawFilledCircle(transformX(x), transformY(y), radius * scale, fill)
        }

        override fun drawCircle(x: Double, y: Double, radius: Double, stroke: AndroidPen?,
                                fill: AndroidPen?) {
            if (fill != null) {
                this@AndroidCanvas.drawFilledCircle(transformX(x), transformY(y), radius * scale, fill)
            }
            if (stroke != null) {
                this@AndroidCanvas.drawCircle(transformX(x), transformY(y), radius * scale, scalePen(stroke)!!)
            }
        }

        override fun drawBitmap(left: Double,
                                top: Double,
                                bitmap: Bitmap,
                                pen: AndroidPen) {
            this@AndroidCanvas.drawBitmap(transformX(left), transformY(top), bitmap, scalePen(pen)!!)
        }

        override fun drawRect(rect: Rectangle, stroke: AndroidPen) {
            this@AndroidCanvas.drawRect(rect.offsetScaled(-xOffset, -yOffset, scale), scalePen(stroke)!!)
        }

        override fun drawFilledRect(rect: Rectangle, fill: AndroidPen) {
            this@AndroidCanvas.drawFilledRect(rect.offsetScaled(-xOffset, -yOffset, scale), scalePen(fill)!!)
        }

        override fun drawRect(rect: Rectangle,
                              stroke: AndroidPen?,
                              fill: AndroidPen?) {
            if (fill != null) {
                this@AndroidCanvas.drawFilledRect(rect.offsetScaled(-xOffset, -yOffset, scale), fill)
            }

            if (stroke != null) {
                this@AndroidCanvas.drawRect(rect.offsetScaled(-xOffset, -yOffset, scale), scalePen(stroke)!!)
            }
        }

        override fun drawPoly(points: DoubleArray,
                              stroke: AndroidPen?,
                              fill: AndroidPen?) {
            this@AndroidCanvas.drawPoly(transform(points), scalePen(stroke), fill)
        }

        private fun transform(points: DoubleArray): DoubleArray {
            val result = DoubleArray(points.size)
            val len = points.size - 1
            var i = 0
            while (i < len) {
                result[i] = transformX(points[i])
                ++i
                result[i] = transformY(points[i])
                ++i
            }
            return result
        }

        fun transformX(x: Double): Double {
            return (x - xOffset) * scale
        }

        fun transformY(y: Double): Double {
            return (y - yOffset) * scale
        }

        override fun drawPath(path: AndroidPath, stroke: AndroidPen?, fill: AndroidPen?) {
            val transformedPath = transformPath(path)
            if (fill != null) {
                this@AndroidCanvas.drawFilledPath(transformedPath, fill.paint)
            }
            if (stroke != null) {
                this@AndroidCanvas.drawPath(transformedPath, scalePen(stroke)!!.paint)
            }
        }

        private fun transformPath(path: AndroidPath): Path {
            val transformedPath = Path(path.path)
            val matrix = Matrix()
            matrix.setScale(scale.toFloat(), scale.toFloat())
            matrix.preTranslate((-xOffset).toFloat(), (-yOffset).toFloat())
            transformedPath.transform(matrix)
            return transformedPath
        }

        override fun drawRoundRect(rect: Rectangle,
                                   rx: Double,
                                   ry: Double,
                                   stroke: AndroidPen) {
            this@AndroidCanvas.drawRoundRect(rect.offsetScaled(-xOffset, -yOffset, scale), rx * scale, ry * scale,
                                             scalePen(
                                                 stroke)!!)
        }

        override fun drawFilledRoundRect(rect: Rectangle,
                                         rx: Double,
                                         ry: Double,
                                         fill: AndroidPen) {
            this@AndroidCanvas.drawFilledRoundRect(rect.offsetScaled(-xOffset, -yOffset, scale), rx * scale,
                                                   ry * scale, scalePen(
                fill)!!)
        }

        override fun drawRoundRect(rect: Rectangle,
                                   rx: Double,
                                   ry: Double,
                                   stroke: AndroidPen?,
                                   fill: AndroidPen?) {
            if (fill != null) {
                this@AndroidCanvas.drawFilledRoundRect(rect.offsetScaled(-xOffset, -yOffset, scale), rx * scale,
                                                       ry * scale, scalePen(fill)!!)
            }
            if (stroke != null) {
                this@AndroidCanvas.drawRoundRect(rect.offsetScaled(-xOffset, -yOffset, scale), rx * scale,
                                                 ry * scale, scalePen(stroke)!!)
            }
        }

        override fun drawText(textPos: Canvas.TextPos,
                              left: Double,
                              baselineY: Double,
                              text: String,
                              foldWidth: Double,
                              pen: AndroidPen) {
            this@AndroidCanvas.drawText(textPos, transformX(left), transformY(baselineY), text, foldWidth * scale,
                                        scalePen(pen)!!, scale)
        }
    }

}
